Foundation/Operation.swift

```diff
@@ -234,81 +234,163 @@ extension OperationQueue {
     public static let defaultMaxConcurrentOperationCount: Int = Int.max
 }
 
+internal class _IndexedOperationLinkedList {
+    
+    internal class Node {
```

Comment:
this is an extraneous allocation; the node contents can be referenced inside the operation themselves.

Reply:
I think you are right. Since both classes, `_IndexedOperationLinkedList` and `_IndexedOperationLinkedList.Node`, are not used out side this file. I think `fileprivate` would be the correct access modifier.

Reply:
Never mind my previous comment, I miss-read that. I read one of your comment below,

> if you flatten from the node into the operations themselves you wont need this and it would reduce execution complexity.

Now I understand. You want an implemented similar to objc as you mentioned in the forum. It would definitely reduce the space complexity if we use `Operation` itself as a node of the linked list. I will updated the implementation and let you know once it is ready for review.

Reply:
@phausler  As I was implement the flatten-approach, I ran into an interesting problem. This property `OperationQueue.operations` is expected to return an array operations in the same order in which all the operations were added.

Given that we need two types of linked list,
**Type 1**: A list to maintain the operations based on the queue priority (`veryLow`, `low`, `normal`, `high` & `veryHigh`)
**Type 2**: A list to maintain the operations based on the order in which it is added (`all`)

So then, the properties `Operation.next` and `Operation.previous` cannot be reused by _Type 2_ as it being used by _Type 1_ linked list.

First I tried using Linked List for all Type 1, and Array for `all`(Type 2). This approach is definitely an improvement, but, this approach didn't provide the performance as I was expecting, as we still running in **O(n)** time complexity.

I liked your approach of using the `Operation` object as a node in linked list, which reduces the space complexity; I don't want to give up that too. Finally I found a solution that works for both, ordered-list and prioritized-list, by added additional two reference in the `Operation` class.

```swift
class Operation {
    /*
      prioritizedPrevious && prioritizedNext used to traverse in the order 
      in which the operations need to be dequeued
    */
    fileprivate weak var prioritizedPrevious: Operation?
    fileprivate var prioritizedNext: Operation?
    
    /*
      orderdedPrevious && orderdedNext used to traverse in the order 
      in which the operations added to the queue
    */
    fileprivate weak var orderdedPrevious: Operation?
    fileprivate var orderdedNext: Operation?
}
```

Let me know what you think.

Reply:
Heres the updated linked list's class looks like,

```swift
fileprivate class _IndexedOperationLinkedList {
    private(set) var root: Operation? = nil
    private(set) var tail: Operation? = nil
    
    func append(_ operation: Operation, inOrder: Bool = false) { }
    private func appendOperationInPriorityList(_ operation: Operation) { }
    private func appendOperationInOrderList(_ operation: Operation) { }
    
    func remove(_ operation: Operation, inOrder: Bool) { }
    private func removeOperationFromPriorityList(_ operation: Operation) { }
    private func removeOperationFromOrderList(_ operation: Operation) { }
    
    func removeFirstInPriority() -> Operation? { }
}

```

Reply:
The operations property is O(n) on Darwin and honestly that is fine since that property is honestly a race condition anyhow and developers should not really use it imho.

Reply:
@phausler  Im couldn't understand your comment, Im kinda getting mixed signal. Do you think the approach I mentioned in the above comments will work or you suggesting some improvements?

Reply:
I’m saying that you should not worry about the performance of the operations property and instead focus on the perf I’d add/remove of singular operations. What you have outlined is roughly the same as what the objc version has modulo weak (the objc side uses unsafe unretained pointers). As long as the operations are strongly referenced by the queue somehow the implementation will be on par.

Reply:
@phausler. I think we are on the same page here. This PR is mainly focused on improving the performance of add/remove single operation. The public property `OperationQueue.operations` caught it the way. I will update the PR addressing the all comments and let you know once its ready for review.

Reply:
Resolving this comment since it is addressed